src/Processors/Formats/Impl/NpyRowInputFormat.cpp

```diff
@@ -0,0 +1,461 @@
+        case NumpyDataTypeIndex::Int32: readBinaryValueAndInsert<T, UInt32>(column->getPtr(), npy_type.getEndianness()); break;
+        case NumpyDataTypeIndex::Int64: readBinaryValueAndInsert<T, UInt64>(column->getPtr(), npy_type.getEndianness()); break;
+        case NumpyDataTypeIndex::UInt8: readBinaryValueAndInsert<T, UInt8>(column->getPtr(), npy_type.getEndianness()); break;
+        case NumpyDataTypeIndex::UInt16: readBinaryValueAndInsert<T, UInt16>(column->getPtr(), npy_type.getEndianness()); break;
+        case NumpyDataTypeIndex::UInt32: readBinaryValueAndInsert<T, UInt32>(column->getPtr(), npy_type.getEndianness()); break;
+        case NumpyDataTypeIndex::UInt64: readBinaryValueAndInsert<T, UInt64>(column->getPtr(), npy_type.getEndianness()); break;
+        default:
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Cannot insert data type into column with type {}",
+                        data_type->getName());
+    }
+}
+
+template <typename T>
+void NpyRowInputFormat::readAndInsertFloat(IColumn * column, const DataTypePtr & data_type, const NumpyDataType & npy_type)
+{
+    switch (npy_type.getTypeIndex())
+    {
+        case NumpyDataTypeIndex::Float32: readBinaryValueAndInsert<T, Float32>(column->getPtr(), npy_type.getEndianness()); break;
+        case NumpyDataTypeIndex::Float64: readBinaryValueAndInsert<T, Float64>(column->getPtr(), npy_type.getEndianness()); break;
+        default:
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Cannot insert data type into column with type {}",
+                        data_type->getName());
+    }
+}
+
+template <typename T>
+void NpyRowInputFormat::readAndInsertString(MutableColumnPtr column, const DataTypePtr & data_type, const NumpyDataType & npy_type, bool is_fixed)
+{
+    size_t size;
+    if (npy_type.getTypeIndex() == NumpyDataTypeIndex::String)
+        size = assert_cast<const NumpyDataTypeString &>(npy_type).getSize();
+    else if (npy_type.getTypeIndex() == NumpyDataTypeIndex::Unicode)
+        size = assert_cast<const NumpyDataTypeUnicode &>(npy_type).getSize();
+    else
+        throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Cannot insert data type into column with type {}",
+                        data_type->getName());
+
+    if (is_fixed)
+    {
+        auto & fixed_string_column = assert_cast<ColumnFixedString &>(*column);
+        size_t n = fixed_string_column.getN();
+        if (size > n)
+            throw Exception(ErrorCodes::TOO_LARGE_STRING_SIZE, "Too large string for FixedString column");
+        fixed_string_column.getChars().resize_fill(fixed_string_column.getChars().size() + n);
+
+        String tmp;
+        tmp.resize(size);
+
+        in->readStrict(tmp.data(), size);
+        tmp.erase(std::remove(tmp.begin(), tmp.end(), '\0'), tmp.end());
+        fixed_string_column.insertData(tmp.c_str(), tmp.size());
```

Comment:
I think you still don't understand this code a bit.
You first resize internal vector of chars to the new size, and then insert new value using `insertData`.  But this method also does resize of `chars` vector:
```
https://github.com/ClickHouse/ClickHouse/blob/93934098fce47a41d07aadf7245c7a1e67a097b8/src/Columns/ColumnFixedString.cpp#L78-L87
```
So, your first resize is useless. Also in FixedString column if value length is less then `N`, it's padded with 0 bytes, so no need to erase them. The best way to insert data into FixedString column in our case is to insert into `chars` directly:
```suggestion
        auto & chars = fixed_string_column.getChars();
        size_t prev_size = chars.size();
        chars.resize_fill(prev_size + n);
        in->readStrict(chars.data() + prev_size, size);
```